### DISCUSSION TITLE
Custom Property Setter vs. `willUpdate` For Retrieving Data Dependent on a Property

### DISCUSSION THREAD

**Discussion by lmahoney1** [lit/lit#4716] [Q&A] [answered]
Hello,

I need to retrieve data through HTTP requests when a given property changes. I have initially implemented the data retrieval in the setter method of the property and it seems to be working well in my scenario.

However, I noticed in the documentation](https://lit.dev/docs/components/properties/#accessors-custom):

> In most cases, you do not need to create custom property accessors. To compute values from existing properties, we recommend using the [willUpdate](https://lit.dev/docs/components/lifecycle/#willupdate) callback, which allows you to set values during the update cycle without triggering an additional update.

I'm wondering if there's a reason `willUpdate` is suggested over custom property accessors?

Thank you!

**Comment by justinfagnani** [comment, marked as answer]
I would use something like Task to retrieve data when properties change: https://lit.dev/docs/data/task/ Task reifies the state of an async operation into synchronous fields that are easily readable in synchronous render methods.

The recommendation to use willUpdate is mainly to synchronize with the reactive update lifecycle. The flow would be that initial reactive properties change, the update lifecycle is triggered, derived data is computed, whether in willUpdate, update, or render, and the render result is created. Using the update lifecycle can help with debouncing and responding to changes to multiple properties at most once. You can definitely use setters if you want though.

**Reply by lmahoney1** [reply]
Thank you very much, I appreciate the insight!
